src/absorbable.rs

```diff
@@ -0,0 +1,267 @@
+            fn to_sponge_bytes(&self) -> Vec<u8> {
+                to_bytes![self].unwrap()
+            }
+
+            fn to_sponge_field_elements(&self) -> Vec<$field<P>> {
+                vec![*self]
+            }
+
+            fn batch_to_sponge_field_elements(batch: &[Self]) -> Vec<$field<P>> {
+                batch.to_vec()
+            }
+        }
+    };
+}
+
+impl_absorbable_field!(Fp256, Fp256Parameters);
+impl_absorbable_field!(Fp320, Fp320Parameters);
+impl_absorbable_field!(Fp384, Fp384Parameters);
+impl_absorbable_field!(Fp768, Fp768Parameters);
+impl_absorbable_field!(Fp832, Fp832Parameters);
+
+macro_rules! impl_absorbable_unsigned {
+    ($t:ident) => {
+        impl<F: PrimeField> Absorbable<F> for $t {
+            fn to_sponge_bytes(&self) -> Vec<u8> {
+                self.to_le_bytes().to_vec()
+            }
+
+            fn to_sponge_field_elements(&self) -> Vec<F> {
+                vec![F::from(*self)]
+            }
+        }
+    };
+}
+
+impl_absorbable_unsigned!(u16);
+impl_absorbable_unsigned!(u32);
+impl_absorbable_unsigned!(u64);
+impl_absorbable_unsigned!(u128);
+
+macro_rules! impl_absorbable_signed {
+    ($signed:ident, $unsigned:ident) => {
+        impl<F: PrimeField> Absorbable<F> for $signed {
+            fn to_sponge_bytes(&self) -> Vec<u8> {
+                self.to_le_bytes().to_vec()
+            }
+
+            fn to_sponge_field_elements(&self) -> Vec<F> {
+                let mut elem = F::from(self.abs() as $unsigned);
+                if *self < 0 {
+                    elem = -elem;
+                }
+                vec![elem]
+            }
+        }
+    };
+}
+
+impl_absorbable_signed!(i8, u8);
+impl_absorbable_signed!(i16, u16);
+impl_absorbable_signed!(i32, u32);
+impl_absorbable_signed!(i64, u64);
+impl_absorbable_signed!(i128, u128);
+
+macro_rules! impl_absorbable_size {
+    ($t:ident) => {
+        impl<F: PrimeField> Absorbable<F> for $t {
+            fn to_sponge_bytes(&self) -> Vec<u8> {
+                Absorbable::<F>::to_sponge_bytes(&(*self as u64))
+            }
+
+            fn to_sponge_field_elements(&self) -> Vec<F> {
+                (*self as u64).to_sponge_field_elements()
+            }
+        }
+    };
+}
+
+impl_absorbable_size!(usize);
+impl_absorbable_size!(isize);
+
+macro_rules! impl_absorbable_group {
+    ($group:ident, $params:ident) => {
+        impl<P: $params, F: PrimeField> Absorbable<F> for $group<P>
+        where
+            P::BaseField: ToConstraintField<F>,
+        {
+            fn to_sponge_bytes(&self) -> Vec<u8> {
+                Absorbable::<F>::to_sponge_bytes(&to_bytes!(self).unwrap())
+            }
+
+            fn to_sponge_field_elements(&self) -> Vec<F> {
+                self.to_field_elements().unwrap()
+            }
+        }
+    };
+}
+
+impl_absorbable_group!(TEAffine, TEModelParameters);
+impl_absorbable_group!(SWAffine, SWModelParameters);
+
+impl<F: PrimeField, A: Absorbable<F>> Absorbable<F> for &[A] {
```

Comment:
If `A` does not have constant size, will this be still secure? `batch_to_sponge_bytes` simply ignores the size, and [`1234`,`56`] and [`12`,`3456`] might be lead to same hash in this way?

Reply:
fixed. If `A` is a struct with variable size, then in (almost) all cases `A` will contain list like structure, and `Absorb` for A will handle this by adding a length